--- session_mysql-1.8/session_mysql.c~	2006-06-15 15:02:32.153398581 +0300
+++ session_mysql-1.8/session_mysql.c	2006-06-15 15:34:15.026147299 +0300
@@ -142,8 +142,24 @@
 	}
 
 	if ((!host && !sock) || !db || !user || !pass) {
+		if (host) {
+			pefree(host, 1);
+		}
+		if (db) {
+			pefree(db, 1);
+		}
+		if (user) {
+			pefree(user, 1);
+		}
+		if (pass) {
+			pefree(pass, 1);
+		}
+		if (sock) {
+			pefree(sock, 1);
+		}
 		return(FAILURE);
 	}
+
 	if (SESSION_MYSQL_G(host)) {
 		pefree(SESSION_MYSQL_G(host),1);
 	}
